### PR TITLE
compiler: Version numbers in SourceFiles

### PR DESCRIPTION
**Pull request by hunger**
We do have version numbers in the `SourceFile`s we generate and pass around for a while now.

Unfortunately those were almost consistently `None`. This PR fixes that.

It is mostly about being able to pass a version number when triggering a compile and about changing the fallback function we have to open files in such a way that it returns the file contents as well as the version.

In all typical cases that version will be `None` (== as on disk), but in the live-preview we can set it to what is in the editor.

An extra patch makes the diagnostics we return contain the version number they refer to. This should (eventually) stop us from showing stale diagnostics.

### CODE REVIEW

File: internal/interpreter/api.rs [hunk 1]
```diff
@@ -576,7 +576,11 @@ impl ComponentCompiler {
         file_loader_fallback: impl Fn(
                 &Path,
             ) -> core::pin::Pin<
-                Box<dyn core::future::Future<Output = Option<std::io::Result<String>>>>,
+                Box<
```
Comment (ogoffart) [inline internal/interpreter/api.rs:579]: This is a breaking change on public API

File: internal/interpreter/api.rs [hunk 2]
```diff
@@ -779,7 +783,11 @@ impl Compiler {
...
```
Comment (ogoffart) [inline internal/interpreter/api.rs:786]: ditto

File: internal/interpreter/api.rs [hunk 3]
```diff
@@ -834,7 +842,28 @@ impl Compiler {
...
+    /// paths while importing.
+    ///
+    /// Any diagnostics produced during the compilation, such as warnings or errors, can be retrieved
+    /// after the call using [`CompilationResult::diagnostics()`].
+    ///
+    /// This function is `async` but in practice, this is only asynchronous if
+    /// [`Self::set_file_loader`] is set and its future is actually asynchronous.
+    /// If that is not used, then it is fine to use a very simple executor, such as the one
+    /// provided by the `spin_on` crate
+    pub async fn build_from_versioned_source(
```
Comment (ogoffart) [inline internal/interpreter/api.rs:860]: this is public API. But SourceFileVersion is not public. I'd rather this to stay internal.

**Review by ogoffart — COMMENTED** [review]
need to be careful when changing public API.

I wonder if we should not provide an internal function to highjack the TypeLoader and have direct access to that. also so that the cache can be re-used.

File: internal/interpreter/api.rs (hunk 1 above)
Comment (hunger) [reply internal/interpreter/api.rs:579]: Any idea on how to avoid it?

Keep this one as is, add a new callback function and use that if set and fall back to the current one?

**Comment by ogoffart** [comment]
> Any idea on how to avoid it?
> Keep this one as is, add a new callback function and use that if set and fall back to the current one?

I think regardless the thing we use it is fine to use the internals from the compiler. I can see a few options:
 - directly change the callback in the CompilerConfig and access the CompilerConfig from the internal API (there is already an accessor to the CompilerConfig)
 - Directly hook into the typeloader somehow. That way we can also speed up things by giving the parsed result¹ directly from the document cache and avoid re-parsing
 - Assume that the preview was build from the document cache anyway: So when we get diagnostics, assume that the file is the one version from the document cache. Since the preview is single threaded, it shouldn't be possible that the cache is updated while the preview is being updated. Or is it possible because of async?

---
¹ I'm just talking about the SyntaxNode here, not necessarily the object_tree's Document, which is harder to share as it gets modified by the passes.

**Comment by hunger** [comment]
> Assume that the preview was build from the document cache anyway: So when we get diagnostics, assume that the file is the one version from the document cache. Since the preview is single threaded, it shouldn't be possible that the cache is updated while the preview is being updated. Or is it possible because of async?

This PR puts the version information into the document cache properly :-)

Right now I have the file version in a separate cache of source code I keep in the live-preview. The source code cache gets updated as soon as a new data gets sent over from the editor/LSP and then triggers the preview to update. While that happens more source code updates can end up in the message queues and will trigger a new update right when the live preview is done rendering. This is what (I think) triggers the problem in #5798.

Async should not effect things too much here: The live-preview gets all contents from a HashMap of stored source code, so all futures we see should be resolvable right away.

I had hoped that VSCode would invalidate all diagnostics that are for outdated source code, but that is apparently not the case. So this will need to do some queue processing to figure out whether the source code was updated and then not send the diagnostics if that is the case.

**Comment by hunger** [comment]
> I'm just talking about the SyntaxNode here, not necessarily the object_tree's Document, which is harder to share as it gets modified by the passes.

Could we change that and make the document/etc. immutable?

**Comment by ogoffart** [comment]
> I had hoped that VSCode would invalidate all diagnostics that are for outdated source code, but that is apparently not the case.

It doesn't. You need to explicitly send an empty diagnostics list for them.

> So this will need to do some queue processing to figure out whether the source code was updated and then not send the diagnostics if that is the case.

We would need to do that also for #5797 and for code that is **not** updated, for all source code dependencies.


> Could we change that and make the document/etc. immutable?

So you mean all the structures from the `object_tree` yes, we could try.

But what the teaching of rust-analyzer is telling us, is that we should actually lazily create the object_tree only for the things we really need (used components).
We can cache the parsing tree (SyntaxNode) but re-create the object_tree on demand.

**Comment by hunger** [comment]
The newest push is restructuring the entire PR...

* It removes the `_versioned_` build function from `ComponentCompiler`: That is internal and unused.
* Adds the function removed above into `Compiler` and adopts the preview to use it. The LSP itself already passes in the version number.
* Changes the fallback function in `ConmponentConfiguration` to return file contents and version, but leaves the setters on `Compiler` unchanged, accepting a function that returns the file contents _only_. This avoids breaking public API as the `ComponentConfiguration` is private.
* Uses all that :-)

**Comment by hunger** [comment]
... and `cargo fmt` the entire thing :-/

File: internal/compiler/diagnostics.rs [hunk 4]
```diff
@@ -285,6 +285,11 @@ impl Diagnostic {
     pub fn source_file(&self) -> Option<&Path> {
         self.span.source_file().map(|sf| sf.path())
     }
+
+    /// return the path of the source file where this error is attached
+    pub fn source_version(&self) -> Option<i32> {
```
Comment (ogoffart) [inline internal/compiler/diagnostics.rs:290]: See the note at the beginning of the `impl` 
This is re-export as public API in slint_interpreter.

Does it really make sense to expose this?

File: internal/compiler/lib.rs [hunk 5]
```diff
@@ -92,7 +92,12 @@ pub struct CompilerConfiguration {
     /// The callback should open the file specified by the given file name and
     /// return an future that provides the text content of the file as output.
     pub open_import_fallback: Option<
-        Rc<dyn Fn(String) -> Pin<Box<dyn Future<Output = Option<std::io::Result<String>>>>>>,
+        Rc<
+            dyn Fn(
+                String,
+            )
+                -> Pin<Box<dyn Future<Output = Option<std::io::Result<(String, Option<i32>)>>>>>,
```
Comment (ogoffart) [inline internal/compiler/lib.rs:99]: We have an alias for that, let's use it.

```suggestion
                -> Pin<Box<dyn Future<Output = Option<std::io::Result<(String, SourceFileVersion)>>>>>,
```

File: internal/interpreter/api.rs [hunk 6]
```diff
@@ -579,8 +579,18 @@ impl ComponentCompiler {
                 Box<dyn core::future::Future<Output = Option<std::io::Result<String>>>>,
             > + 'static,
     ) {
-        self.config.open_import_fallback =
-            Some(Rc::new(move |path| file_loader_fallback(Path::new(path.as_str()))));
+        let file_loader_fallback = Rc::new(file_loader_fallback);
```
Comment (ogoffart) [inline internal/interpreter/api.rs:582]: This looks like a useless allocation. Why is this needed?

File: internal/interpreter/api.rs [hunk 7]
```diff
@@ -579,8 +579,18 @@ impl ComponentCompiler {
...
+
+        let open_import_fallback = move |file_name: String| -> core::pin::Pin<
+            Box<dyn core::future::Future<Output = Option<std::io::Result<(String, Option<i32>)>>>>,
+        > {
+            let flfb = file_loader_fallback.clone();
+            Box::pin({
+                async move { flfb(&PathBuf::from(file_name)).await.map(|r| r.map(|c| (c, None))) }
+            })
+        };
```
Comment (ogoffart) [inline internal/interpreter/api.rs:591]: I think we can manage without putting the file_loader_fallback in a extra RC thereby saving one allocation and one indirection:

```suggestion
        let open_import_fallback = move |file_name: String| -> core::pin::Pin<
            Box<dyn core::future::Future<Output = Option<std::io::Result<(String, Option<i32>)>>>>,
        > {
            let flfb = file_loader_fallback(&Path::new(file_name.as_str()));
            Box::pin(async move { flfb.await.map(|r| r.map(|c| (c, None))) })
        };
```

(untested)

File: internal/interpreter/api.rs [hunk 8]
```diff
@@ -782,8 +754,18 @@ impl Compiler {
                 Box<dyn core::future::Future<Output = Option<std::io::Result<String>>>>,
             > + 'static,
     ) {
-        self.config.open_import_fallback =
-            Some(Rc::new(move |path| file_loader_fallback(Path::new(path.as_str()))));
+        let file_loader_fallback = Rc::new(file_loader_fallback);
```
Comment (ogoffart) [inline internal/interpreter/api.rs:757]: ditto

File: tools/lsp/common.rs [hunk 9]
```diff
@@ -677,7 +677,7 @@ pub enum PreviewToLspMessage {
     /// Show a status message in the editor
     Status { message: String, health: crate::lsp_ext::Health },
     /// Report diagnostics to editor.
-    Diagnostics { uri: Url, diagnostics: Vec<lsp_types::Diagnostic> },
+    Diagnostics { uri: Url, version: UrlVersion, diagnostics: Vec<lsp_types::Diagnostic> },
```
Comment (ogoffart) [inline tools/lsp/common.rs:680]: btw, what's the difference between `SourceFileVersion` ans `UrlVersion`? Should we not get rid of `UrlVersion` in favor of `SourceFileVersion` everywhere?

File: tools/slintpad/src/editor_widget.ts [hunk 10]
```diff
@@ -551,37 +551,38 @@ class EditorPaneWidget extends Widget {
     protected async handle_lsp_url_request(
         era: number,
         url: string,
-    ): Promise<string> {
+    ): Promise<[string, null | number]> {
         if (this.#url_mapper === null) {
-            return Promise.resolve("Error: Can not resolve URL.");
+            return Promise.resolve(["Error: Can not resolve URL.", null]);
```
Comment (ogoffart) [inline tools/slintpad/src/editor_widget.ts:556]: So we resolve it, but it's an error? Seems wrong to me.

**Review by ogoffart — COMMENTED** [review]
Did you check that it actually fixes #5798?

File: internal/compiler/diagnostics.rs (hunk 4 above)
Comment (hunger) [reply internal/compiler/diagnostics.rs:290]: How would you extract that info without this helper? Everything in that struct is private, so I think I will need an accessor to get the information out.

It should return a SourceFileVersion though.

File: internal/interpreter/api.rs (hunk 7 above)
Comment (hunger) [reply internal/interpreter/api.rs:591]: You are right! Wonderful.

File: tools/lsp/common.rs (hunk 9 above)
Comment (hunger) [reply tools/lsp/common.rs]: That is already merged as part of #5842.

File: tools/slintpad/src/editor_widget.ts (hunk 10 above)
Comment (hunger) [reply tools/slintpad/src/editor_widget.ts:556]: Still open :-)

Comment (hunger) [reply tools/slintpad/src/editor_widget.ts:556]: Done now.

File: internal/compiler/diagnostics.rs [hunk 11]
```diff
@@ -285,6 +285,11 @@ impl Diagnostic {
     pub fn source_file(&self) -> Option<&Path> {
         self.span.source_file().map(|sf| sf.path())
     }
+
+    /// return the path of the source file where this error is attached
+    pub fn source_version(&self) -> SourceFileVersion {
```
Comment (ogoffart) [inline internal/compiler/diagnostics.rs:290]: I'm still uncomfortable having this in the public API. Can it somehow be gated? Or what do you think, @tronical

Comment (tronical) [reply internal/compiler/diagnostics.rs:290]: Same here. The docs say that this returns a path, but it returns an `Option<i32>` and it's not clear at all what the use of this is.

**Comment by tronical** [comment]
If the objective is to get version numbers more or less passed through so that diagnostics can be associated with correct versions, then from the compiler point of view that seems useless to pass it. Would it perhaps be possible to do this on the caller side?

1. Record a list of all files and their versions.
2. Feed the source code to compile into the compiler.
3. In the callback when sources are requested, extend the "book keeping" data structure of step 1 accordingly.
4. When a diagnostic is reported / extracted afterwards, determine the corresponding version via lookup in the book keeping data structure of step 1.

**Comment by hunger** [comment]
Sure, that can be done.

We moved from having such a data structure to just storing the version number in the `SourceFileInner`. This PR adds one accessor to that information -- which I admit is probably stupid as nobody can actually put the version in place without internal API.

Stabilizing the version handling would require us to change the fallback file loader in the `CompilerConfiguration`, which as @ogoffart has pointed out is a API break. So should we remove the entire thing again and I'll work around the missing information in the LSP?

**Comment by ogoffart** [comment]
Why can we not get the version from the TypeLoader?

**Comment by hunger** [comment]
@ogoffart: Are you suggesting to keep the actual version information in the `SourceFileInner` and access it trough the type loader from there or to store the version next to the document directly in the type loader?

**Comment by hunger** [comment]
Nevermind: I'll remove the version getter from the diagnosticas and use the LSP's DocumentCache instead.

File: internal/compiler/diagnostics.rs (hunk 11 above)
Comment (hunger) [reply internal/compiler/diagnostics.rs]: It's gone now.

File: internal/interpreter/api.rs [hunk 12]
```diff
@@ -578,8 +579,18 @@ impl ComponentCompiler {
...
     ) {
-        self.config.open_import_fallback =
-            Some(Rc::new(move |path| file_loader_fallback(Path::new(path.as_str()))));
+        let open_import_fallback = move |file_name: String| -> core::pin::Pin<
+            Box<
+                dyn core::future::Future<
+                    Output = Option<std::io::Result<(String, SourceFileVersion)>>,
+                >,
+            >,
+        > {
```
Comment (ogoffart) [inline internal/interpreter/api.rs:588]: I wonder if this works:
```suggestion
        let open_import_fallback = move |file_name: String| {
```

Or at least use `_`  or `use` to make this beast readable.

File: tools/slintpad/src/lsp.ts [hunk 13]
```diff
@@ -56,7 +56,7 @@ function createLanguageClient(
     return client;
 }
 
-export type FileReader = (_url: string) => Promise<string>;
+export type FileReader = (_url: string) => Promise<[string, null | number]>;
```
Comment (ogoffart) [inline tools/slintpad/src/lsp.ts:59]: i wonder if it wouldn't be clean to add an named alias for `[string, null|number]` with a comment explaining the meaning of the number.

**Review by ogoffart — APPROVED** [review]
Did you check that it actually fixes https://github.com/slint-ui/slint/issues/5798?

If it does fix it, then this looks good now.

File: tools/slintpad/src/lsp.ts (hunk 13 above)
Comment (hunger) [reply tools/slintpad/src/lsp.ts]: Good idea. I added a patch exporting those types from Rust to WASM/TS and using them wheer appropriate.

The LSP protocol uses `null|integer` for this, so there is no nice named type available there.

**Comment by hunger** [comment]
Let's remove the entire version information thing from the compiler and do all of it in the LSP.

#5875 has the PR for that.
